x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/oidc/TransportOpenIdConnectLogoutAction.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.security.action.oidc;
+
+import com.nimbusds.jwt.JWT;
+import com.nimbusds.jwt.JWTParser;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.ElasticsearchSecurityException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.security.action.oidc.OpenIdConnectLogoutAction;
+import org.elasticsearch.xpack.core.security.action.oidc.OpenIdConnectLogoutRequest;
+import org.elasticsearch.xpack.core.security.action.oidc.OpenIdConnectLogoutResponse;
+import org.elasticsearch.xpack.core.security.authc.Authentication;
+import org.elasticsearch.xpack.core.security.authc.Realm;
+import org.elasticsearch.xpack.core.security.authc.support.TokensInvalidationResult;
+import org.elasticsearch.xpack.core.security.user.User;
+import org.elasticsearch.xpack.security.authc.Realms;
+import org.elasticsearch.xpack.security.authc.TokenService;
+import org.elasticsearch.xpack.security.authc.oidc.OpenIdConnectRealm;
+
+import java.io.IOException;
+import java.text.ParseException;
+import java.util.Map;
+
+/**
+ * Transport action responsible for generating an OpenID connect logout request to be sent to an OpenID Connect Provider
+ */
+public class TransportOpenIdConnectLogoutAction extends HandledTransportAction<OpenIdConnectLogoutRequest, OpenIdConnectLogoutResponse> {
+
+    private final Realms realms;
+    private final TokenService tokenService;
+    private static final Logger logger = LogManager.getLogger(TransportOpenIdConnectLogoutAction.class);
+
+    @Inject
+    public TransportOpenIdConnectLogoutAction(TransportService transportService, ActionFilters actionFilters, Realms realms,
+                                              TokenService tokenService) {
+        super(OpenIdConnectLogoutAction.NAME, transportService, actionFilters,
+            (Writeable.Reader<OpenIdConnectLogoutRequest>) OpenIdConnectLogoutRequest::new);
+        this.realms = realms;
+        this.tokenService = tokenService;
+    }
+
+    @Override
+    protected void doExecute(Task task, OpenIdConnectLogoutRequest request, ActionListener<OpenIdConnectLogoutResponse> listener) {
+        invalidateRefreshToken(request.getRefreshToken(), ActionListener.wrap(ignore -> {
```

Comment:
I would leave this last, after we make sure the token is valid and the authn token has been invalidated. This way the `refresh_token` is usable even when this API returns an error (validation or otherwise).
Do you think otherwise?

Reply:
The idea behind invalidating the refresh token first (both here and in the SAML logout , or the `invalidate` API) is that since our refresh tokens and access tokens are linked, invalidation usually means both tokens. Now if we invalidate the access token _before_ the refresh token, theoretically we allow for an extra window (while we invalidate the access token) where the refresh token can be used to get a new refresh token, thus persisting access.

Reply:
I see, it makes sense, thanks.

I think, it would be better to address the root problem, to not allow _refreshing invalidated access tokens_ or that _invalidating the refresh token would also invalidate the access token_.

Should probably be done in a follow-up, WDYT?

Reply:
> to not allow _refreshing invalidated access tokens_

This is not the root problem. The root problem here is that we are "logging out a user" and that user has both an access and a refresh token. Since the refresh token can be used to get a new refresh token (and access token) , and since we use this step invalidation process, it makes sense that the refresh token should be invalidated first. 

> or that _invalidating the refresh token would also invalidate the access token_.

We don't _necessarily_ want to do this. There is a valid use case where a client might want to invalidate the refresh token so that the user can only use their current access token for the remaining duration of it but not be able to refresh. 

I think there is value in your suggestion and we could add an `invalidatedTokenDocument` method in the TokenService and call that from the various TransportRealmLogoutAction's. I'll address this in a follow up.

Reply:
> There is a valid use case where a client might want to invalidate the refresh token so that the user can only use their current access token for the remaining duration of it but not be able to refresh.

Sure we can be flexible here, but the client should not _act like he owns it_ . The authorization server can invalidate tokens even without the client calling an API (to mitigate threats, etc) I am leaning towards this being the wrong kind of flexibility because it doesn't bring any benefit to the client.

> I think there is value in your suggestion and we could add an `invalidatedTokenDocument` method in the TokenService and call that from the various TransportRealmLogoutAction's.

I am 👍 to this, but see below:

I flagged this as a problem because it makes things "smart". I understand the reasoning from a protocol point of view, but I don't like it as a programmer that these functions _should_ be called in order if they don't pass a return value from one to the other.

As a data point, the RFC https://tools.ietf.org/html/rfc7009#section-2.1 says:

> Depending on the authorization server's revocation policy, the
   revocation of a particular token may cause the revocation of related
   tokens and the underlying authorization grant.  If the particular
   token is a refresh token and the authorization server supports the
   revocation of access tokens, then the authorization server SHOULD
   also invalidate all access tokens based on the same authorization
   grant (see Implementation Note).  If the token passed to the request
   is an access token, the server MAY revoke the respective refresh
   token as well.

I think we should go with SHOULD and invalidate the access token when the refresh token is invalidated.

Reply:
We're digressing but I'm enjoying the discussion :) 

We had gone back and forth on this also when the refresh functionality was introduced back in January. I don't have a very clear recollection on whether we consciously decided to not invalidate access tokens when invalidating refresh tokens, I do remember we consciously decided to not invalidate refresh tokens when invalidating the corresponding access token. 

I think your RTFRFC persuaded me, this could also simplify the TokenService code further. I'll be taking it up as a follow up task and we can iterate on that PR more if needed